File: benchmarks/bench_frame_parse.cpp
// bench_frame_parse.cpp — parse/build throughput for 1 MiB binary payload

#include "bench_common.hpp"
#include <wscpp/frame/parser.hpp>
#include <wscpp/frame/builder.hpp>
#include <cstdint>
#include <cstdio>
#include <vector>

using namespace wscpp;
using namespace wscpp::bench;

int main() {
    const std::size_t payload_size = 1024 * 1024;
    const int iterations = 50;

    std::vector<uint8_t> payload(payload_size);
    for (std::size_t i = 0; i < payload_size; ++i) {
        payload[i] = static_cast<uint8_t>(i & 0xFF);
    }

    frame::builder builder;
    frame::parser parser;
    frame::frame_header header;
    std::vector<uint8_t> parsed;

    const clock::time_point build_start = clock::now();
    std::vector<uint8_t> frame;
    for (int i = 0; i < iterations; ++i) {
        frame = builder.build_binary(payload.data(), payload.size(), true, false);
    }
    const clock::time_point build_end = clock::now();

    const double build_sec = elapsed_sec(build_start, build_end);
    print_throughput("frame_build", static_cast<double>(payload_size * iterations),
                     build_sec, iterations);

    const clock::time_point parse_start = clock::now();
    for (int i = 0; i < iterations; ++i) {
        parser.reset();
        const frame::parse_result result =
            parser.parse(frame.data(), frame.size(), header, parsed);
        if (result != frame::parse_result::COMPLETE || parsed.size() != payload_size) {
            std::fprintf(stderr, "parse failed\n");
            return 1;
        }
    }
    const clock::time_point parse_end = clock::now();

    const double parse_sec = elapsed_sec(parse_start, parse_end);
    print_throughput("frame_parse", static_cast<double>(payload_size * iterations),
                     parse_sec, iterations);

    return 0;
}

File: rust/ws-rs-benches/src/bin/bench_frame_parse.rs
// bench_frame_parse — parse/build throughput for 1 MiB binary payload (mirrors C++ bench_frame_parse)

use std::time::Instant;

use ws_rs::frame::{FrameBuilder, ParseResult, Parser};
use ws_rs_benches::common::print_throughput;

fn main() {
    const PAYLOAD_SIZE: usize = 1024 * 1024;
    const ITERATIONS: i32 = 50;

    let mut payload = vec![0u8; PAYLOAD_SIZE];
    for (i, b) in payload.iter_mut().enumerate() {
        *b = (i & 0xFF) as u8;
    }

    let build_start = Instant::now();
    let mut frame = Vec::new();
    for _ in 0..ITERATIONS {
        frame = FrameBuilder::build_binary(&payload, true, false);
    }
    let build_sec = ws_rs_benches::common::elapsed_sec(build_start, Instant::now());
    print_throughput(
        "frame_build",
        PAYLOAD_SIZE as f64 * ITERATIONS as f64,
        build_sec,
        ITERATIONS,
    );

    let parse_start = Instant::now();
    for _ in 0..ITERATIONS {
        let mut parser = Parser::new();
        let (result, _) = parser.parse(&frame);
        if result != ParseResult::Complete || parser.frame().payload.len() != PAYLOAD_SIZE {
            eprintln!("parse failed");
            std::process::exit(1);
        }
    }
    let parse_sec = ws_rs_benches::common::elapsed_sec(parse_start, Instant::now());
    print_throughput(
        "frame_parse",
        PAYLOAD_SIZE as f64 * ITERATIONS as f64,
        parse_sec,
        ITERATIONS,
    );
}
